src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.cs

```diff
@@ -348,12 +355,19 @@ public IEnumerable<ILSequencePoint> GetSequencePointsForMethod(MethodDesc method
             return moduleData.PdbReader.GetSequencePointsForMethod(MetadataTokens.GetToken(ecmaMethod.Handle));
         }
 
-        public IEnumerable<ILLocalVariable> GetLocalVariableNamesForMethod(MethodDesc method)
+        public IEnumerable<ILLocalVariable> GetLocalVariableNamesForMethod(MethodDesc method, MethodIL methodIL)
         {
-            EcmaMethod ecmaMethod = method.GetTypicalMethodDefinition() as EcmaMethod;
+            MethodDesc uninstantiatedMethod = method.GetTypicalMethodDefinition();
+            MethodIL uninstantiatedMethodIL = methodIL.GetMethodILDefinition();
+
+            EcmaMethod ecmaMethod = uninstantiatedMethod as EcmaMethod;
             if (ecmaMethod == null)
                 return null;
 
+            // If IL is not backed by EcmaMethodIL, the body was generated by the compiler and the PDB won't match.
+            if (!(uninstantiatedMethodIL is EcmaMethodIL))
```

Comment:
This assumes that when we always substitute EcmaMethodIL with non-EcmaMethodIL. It happens to be the case today, but it would be nice if we handled substituting one EcmaMethodIL with a different EcmaMethodIL (we are not that far from it with the MCG interop hook ...).

Would it make sense to change the return type of the new virtual method to MethodDesc (or maybe even Object)? It would identify where the IL for this method actually came from.